File: src/Modelo/Disciplina.java
package Modelo;

import java.io.Serializable;
import java.util.*;

public class Disciplina implements Serializable, Comparable<Disciplina>{

    private static final long serialVersionUID = 1L;
    private String codigo;
    private String nome;
    private Periodo periodo;
    private Docente professor;
    private ArrayList<Atividade> atividades;

    public Disciplina(String codigo, String nome, Periodo periodo, Docente professor) {
        this.codigo = codigo;
        this.nome = nome;
        this.periodo = periodo;
        this.professor = professor;
        atividades = new ArrayList<>();
    }

    public String obterCodigo() {
        return codigo;
    }

    public String obterNome() {
        return nome;
    }

    public Periodo obterPeriodo() {
        return periodo;
    }

    public Docente obterDocente() {
        return professor;
    }

    public void anexaAtividade(Atividade atividade){
        atividades.add(atividade);
    }

    public ArrayList<Atividade> obterAtividades(){
        return atividades;
    }

    public double obterNotaTotal(){
        double notaTotal = 0;
        for(Atividade ativ : atividades){
            notaTotal += ativ.calculaNota();
        }
        return notaTotal;
    }

    public int obterAvaliacoesTotal(){
        int qtdAvaliacao = 0;
        for(Atividade ativ : atividades){
            qtdAvaliacao += ativ.obterAvaliacoes().size();
        }
        return qtdAvaliacao;
    }

    public static int comparePeriodo(Disciplina disc1, Disciplina disc2){
        if( disc1.periodo.obterCodigo().compareTo(disc2.periodo.obterCodigo()) == 0){
            return disc1.codigo.compareToIgnoreCase(disc2.codigo);
        }
        return disc1.periodo.obterCodigo().compareTo(disc2.periodo.obterCodigo());
    }

    @Override
    public int compareTo(Disciplina disc2){
        return nome.compareToIgnoreCase(disc2.nome);
    }

    @Override
    public String toString() {
        return "Codigo: "+ codigo+"-"+periodo.obterAno()+ "/" +periodo.obterSemestre() +" - Nome: "+ nome + 
        " - Professor: "+ professor.obterNome();
    }

    public int obterSincronas(){
        int qtdSinc = 0;
        for(Atividade a : obterAtividades()){
            if(a.obterSincrona()){
                qtdSinc++;
            }
        }
        return qtdSinc;
    }

    public int obterAvalitiva(){
        int qtdAvaliativas = 0;
        for(Atividade a : obterAtividades()){
            if(a.ehAvaliativa()) {
                qtdAvaliativas++;
            }
        }
        return qtdAvaliativas;
    }

    public int obterCargaHoraria(){
        int cargaHoraria = 0;
        for(Atividade a : obterAtividades()){
            cargaHoraria += a.obterCargaHoraria();
        }
        return cargaHoraria;
    }
}

File: cpp_version/Modelo/Disciplina.cpp
#include "Disciplina.h"

Disciplina :: Disciplina(string codigo, string nome, Periodo *periodo, Docente *professor) {
    this->codigo = codigo;
    this->nome = nome;
    this->periodo = periodo;
    this->professor = professor;
}

Disciplina :: ~Disciplina(){
    delete periodo;
    delete professor;
}

string Disciplina :: obterCodigo() const{
    return codigo;
}

string Disciplina :: obterNome() const{
    return nome;
}

Periodo* Disciplina :: obterPeriodo() const{
    return periodo;
}

Docente* Disciplina :: obterDocente() const{
    return professor;
}

vector<Atividade*> Disciplina :: obterAtividades(){
    return atividades;
}

void Disciplina :: anexaAtividade(Atividade* atividade){
    atividades.push_back(atividade);
}

double Disciplina :: obterNotaTotal(){
    double notaTotal = 0;
    for(Atividade* ativ : atividades){
        notaTotal += ativ->obterAvaliacoes().size();
    }
    return notaTotal;
}

int Disciplina :: obterAvaliacoesTotal(){
    int qtdAvaliacao = 0;
    for(Atividade* ativ : atividades){
        qtdAvaliacao += ativ->obterAvaliacoes().size();
    }
    return qtdAvaliacao;
}

int Disciplina :: obterSincronas(){
    int qtdSinc = 0;
    for(Atividade* a : obterAtividades()){
        if(a->obterSincrona()){
            qtdSinc++;
        }
    }
    return qtdSinc;
}

int Disciplina :: obterAvalitiva(){
    int qtdAvaliativas = 0;
    for(Atividade* a : obterAtividades()){
        if(a->ehAvaliativa()) {
            qtdAvaliativas++;
        }
    }
    return qtdAvaliativas;
}

int Disciplina :: obterCargaHoraria(){
    int cargaHoraria = 0;
    for(Atividade* a : obterAtividades()){
        cargaHoraria += a->obterCargaHoraria();
    }
    return cargaHoraria;
}

double Disciplina :: percentualAtividadesSincronas(){
    double cont=0, cont2=0;

    if(atividades.size() == 0) return 0;

    for(Atividade* a : obterAtividades()){
        cont2++;
        if(a->obterSincrona())
            cont ++;
    }

    return (cont/cont2)*100;
}

double Disciplina :: percentualAtividadesAssincronas(){
    double cont=0, cont2=0;

    if(atividades.size() == 0) return 0;

    for(Atividade* a : obterAtividades()){
        cont2++;
        if(!a->obterSincrona())
            cont ++;
    }

    return (cont/cont2)*100;
}

int Disciplina :: calculaCargaHorariaTotal(){
    double cont=0;

    for(Atividade* a : obterAtividades()){
        cont += a->obterCargaHoraria();
    }

    return cont;
}

int Disciplina :: obterQtdAvaliacoesTotal(){
    int cont=0;

    for(Atividade* a : obterAtividades()){
        cont += a->obterQtdAvaliacoes();
    }

    return cont;
}

bool Disciplina :: comparaNome(const Disciplina* esq, const Disciplina* dir){
    return stringCompare(esq->obterNome(), dir->obterNome());
}

bool Disciplina :: compara(const Disciplina* esq, const Disciplina* dir){
    int cmp = stringCompare(esq->obterPeriodo()->obterCodigo(), dir->obterPeriodo()->obterCodigo());
    if(cmp != 0) return cmp;
    return stringCompare(esq->obterCodigo(), dir->obterCodigo());
}

ostream& operator<<(ostream& os, Disciplina &disciplina){
    os << "Codigo: " << disciplina.codigo << "-" << disciplina.periodo->obeterAno() << "/" << disciplina.periodo->obterSemestre() << " - Nome: " << disciplina.nome << " - Professor: " << disciplina.professor->obterNome() << endl;
    return os;
}
